### ISSUE TITLE
Understanding the cutoff ensemble calculation

### ISSUE THREAD

**Issue by jacksonwr5798** [biomodhub/biomod2#506]
I was wondering if you could provide further explanation of the calculation for the cutoff values generated for the ensemble models. I saw the explanation in a previous issue for the calculation of the cutoff values for each individual model, but it does not explain how the value is calculated for the ensemble model. I have attached csv files with the output for each individual model as well as the cutoff values for the ensemble. Based on the kept models, I would have expected the cutoff values for the ensemble to higher. Any information or guidance on this topic would be greatly appreciated.

I am happy to provide further data, code, and output if that would be helpful to answering my question.

Thanks,
Jackson 
[cutoff_20240919.csv](https://github.com/user-attachments/files/17067158/cutoff_20240919.csv)
[cutoff_ensemble_20240916.csv](https://github.com/user-attachments/files/17067159/cutoff_ensemble_20240916.csv)
[ensemble_kept_20240919.csv](https://github.com/user-attachments/files/17067160/ensemble_kept_20240919.csv)

```
 Amphi_data <- BIOMOD_FormatingData(
    resp.var = Amphi_occ['Anaxyrus.quercicus'], #presence data and update for new species ---
    resp.xy = Amphi_occ[, c('longitude', 'latitude')], #location data
    expl.var = bioclim_SE_US_sub, #climate variables
    resp.name = "Anaxyrus.quercicus", ##update when looking at different species ---
    PA.nb.rep = 11, #number of sets of pseudo absences selected
    PA.nb.absences = c(rep(790,10), rep(10000,1)), #, rep(1000,10)#number pseudo absences selected --> match number of obs. from thinned data
    PA.strategy = 'random' #heard that the random/disk approaches are the best
    )
Amphi_data

allModels <- c( 'FDA', 'GAM', 'GBM', 'GLM',
                'MARS', 'MAXNET', 'RF', 'XGBOOST', 'MAXENT', 
                'ANN', 'CTA', 'SRE')
SelectModels <- c('GBM','XGBOOST', 'GLM')

#load default model options to tune the selected models
Amphi_Options <- bm_ModelingOptions(
  data.type = "binary",
  models = allModels, #SelectModels,
  strategy = "default",
  user.val = NULL,
  user.base = "bigboss",
  bm.format = Amphi_data,
  calib.lines = NULL)
Amphi_Options

#Tune GBM
tuned.gbm <- bm_Tuning(model = 'GBM',
                       tuning.fun = 'gbm', ## see in ModelsTable
                       do.formula = FALSE,
                       #do.stepAIC = FALSE,
                       bm.options = Amphi_Options@options$GBM.binary.gbm,
                       metric.eval = "ROC",
                       metric.AIC = "AIC",
                       bm.format = Amphi_data)
tuned.gbm

#Tune xgboost
tuned.xgboost <- bm_Tuning(model = 'XGBOOST',
                                    tuning.fun = 'xgbTree', ## see in ModelsTable
                                    do.formula = FALSE,
                                    #do.stepAIC = FALSE,
                                    bm.options = Amphi_Options@options$XGBOOST.binary.xgboost,
                                    metric.eval = "ROC",
                                    metric.AIC = "AIC",
                                    bm.format = Amphi_data)

tuned.xgboost

#updated 8/5/24
#Define user settings based on tuning output for each run
user.GBM <- list('_PA1_allRun' = list(n.trees = 1000,
                                 interaction.depth = 8,
                                 n.minobsinnode = 10,
                                 shrinkage = 0.01),
                 '_PA2_allRun' = list(n.trees = 500,
                                 interaction.depth = 8,
                                 n.minobsinnode = 10,
                                 shrinkage = 0.01),
                 '_PA3_allRun' = list(n.trees = 1000,
                                 interaction.depth = 8,
                                 n.minobsinnode = 10,
                                 shrinkage = 0.01),
                 '_PA4_allRun' = list(n.trees = 2500,
                                 interaction.depth = 8,
                                 n.minobsinnode = 10,
                                 shrinkage = 0.01),
                 '_PA5_allRun' = list(n.trees = 2500,
                                 interaction.depth = 8,
                                 n.minobsinnode = 10,
                                 shrinkage = 0.01),
                 '_PA6_allRun' = list(n.trees = 1000,
                                 interaction.depth = 8,
                                 n.minobsinnode = 10,
                                 shrinkage = 0.01),
                 '_PA7_allRun' = list(n.trees = 500,
                                 interaction.depth = 8,
                                 n.minobsinnode = 10,
                                 shrinkage = 0.001),
                 '_PA8_allRun' = list(n.trees = 500,
                                 interaction.depth = 8,
                                 n.minobsinnode = 10,
                                 shrinkage = 0.01),
                 '_PA9_allRun' = list(n.trees = 500,
                                 interaction.depth = 5,
                                 n.minobsinnode = 10,
                                 shrinkage = 0.01),
                 '_PA10_allRun' = list(n.trees = 500,
                                  interaction.depth = 8,
                                  n.minobsinnode = 10,
                                  shrinkage = 0.001))#,
# '_allData_allRun' = list(n.trees = 2500,
  #                     interaction.depth = 7,
  ##                   n.minobsinnode = 5,
  #                 shrinkage = 0.001))

#updated 8/5/24
user.XGBOOST <- list('_PA1_allRun' = list(nrounds = 50,
                                     max_depth = 1,
                                     eta = 0.3,
                                     gamma = 0,
                                     colsample_bytree = 0.8,
                                     min_child_weight = 1,
                                     subsample = 0.5),
                     '_PA2_allRun' = list(nrounds = 50,
                                     max_depth = 1,
                                     eta = 0.4,
                                     gamma = 0,
                                     colsample_bytree = 0.8,
                                     min_child_weight = 1,
                                     subsample = 0.5),
                     '_PA3_allRun' = list(nrounds = 50,
                                     max_depth = 1,
                                     eta = 0.3,
                                     gamma = 0,
                                     colsample_bytree = 0.8,
                                     min_child_weight = 1,
                                     subsample = 0.5),
                     '_PA4_allRun' = list(nrounds = 50,
                                     max_depth = 1,
                                     eta = 0.4,
                                     gamma = 0,
                                     colsample_bytree = 0.8,
                                     min_child_weight = 1,
                                     subsample = 0.5),
                      '_PA5_allRun' = list(nrounds = 50,
                                           max_depth = 1,
                                           eta = 0.3,
                                           gamma = 0,
                                           colsample_bytree = 0.8,
                                           min_child_weight = 1,
                                           subsample = 0.5),
                      '_PA6_allRun' = list(nrounds = 50,
                                           max_depth = 1,
                                           eta = 0.4,
                                           gamma = 0,
                                           colsample_bytree = 0.8,
                                           min_child_weight = 1,
                                           subsample = 0.5),
                       '_PA7_allRun' = list(nrounds = 50,
                                       max_depth = 1,
                                       eta = 0.3,
                                       gamma = 0,
                                       colsample_bytree = 0.8,
                                       min_child_weight = 1,
                                       subsample = 0.5),
                        '_PA8_allRun' = list(nrounds = 50,
                                       max_depth = 1,
                                       eta = 0.3,
                                       gamma = 0,
                                       colsample_bytree = 0.6,
                                       min_child_weight = 1,
                                       subsample = 0.5),
                         '_PA9_allRun' = list(nrounds = 50,
                                       max_depth = 1,
                                         eta = 0.3,
                                         gamma = 0,
                                         colsample_bytree = 0.8,
                                         min_child_weight = 1,
                                         subsample = 0.5),
                          '_PA10_allRun' = list(nrounds = 50,
                                           max_depth = 1,
                                           eta = 0.4,
                                           gamma = 0,
                                           colsample_bytree = 0.8,
                                           min_child_weight = 1,
                                           subsample = 0.5))

user.GLM <- list('_PA11_allRun' = list())#family = binomial(link = 'logit'),

#compile user validated model settings
user.val <- list(GBM.binary.gbm.gbm = user.GBM,
                 XGBOOST.binary.xgboost.xgboost = user.XGBOOST,
                 GLM.binary.stats.glm = user.GLM)

#Update model options with user defined model settings
Amphi_Options <- bm_ModelingOptions(data.type = 'binary',
                                    models = c( "GBM", "XGBOOST", "GLM"),
                                    strategy = "user.defined",
                                    user.val = user.val,
                                    #user.base = "bigboss",
                                    bm.format = Amphi_data)
Amphi_Options




AmphiModelOut <- BIOMOD_Modeling(bm.format = Amphi_data,
                                             modeling.id = 'AllModels',
                                             models = c("GLM",
                                                        "GBM",
                                                        #"RF", #select final algo options based on ROC/TSS plot
                                                        #"GAM",
                                                        #"CTA",
                                                        #"ANN", #ANN models fail every time
                                                        #"FDA",
                                                        #"SRE",
                                                        #"MARS",
                                                        #"MAXENT", 
                                                        #"MAXNET",
                                                        "XGBOOST"
                                                        ),
                                              bm.options = Amphi_Options, #model settings --> select tuned or default settings above
                                              models.pa = list(
                                                 "GBM" = c("PA1","PA2","PA3","PA4","PA5","PA6","PA7","PA8","PA9","PA10"),
                                                 "XGBOOST" = c("PA1","PA2","PA3","PA4","PA5","PA6","PA7","PA8","PA9","PA10"),
                                                 #"ANN" = c("PA1","PA2","PA3","PA4","PA5","PA6","PA7","PA8","PA9","PA10"),
                                                 #"RF" = c("PA1","PA2","PA3","PA4","PA5","PA6","PA7","PA8","PA9","PA10"),
                                                 #"CTA" = c("PA1","PA2","PA3","PA4","PA5","PA6","PA7","PA8","PA9","PA10"),
                                                 #"FDA" = c("PA11","PA12","PA13","PA14","PA15","PA16","PA17","PA18","PA19","PA20"),
                                                 #"SRE" = c("PA11","PA12","PA13","PA14","PA15","PA16","PA17","PA18","PA19","PA20"),
                                                 #"MARS" = c("PA11","PA12","PA13","PA14","PA15","PA16","PA17","PA18","PA19","PA20"),
                                                 "GLM" = c("PA11")), 
                                                 #"GAM" = c("PA21"),
                                                 #"MAXNET" = c("PA21")),
                                                 #"MAXENT" = c("PA21")), #Removed MAXENT because it was lowering the cutoff significantly
                                               CV.strategy = 'env',
                                               CV.k = 2,
                                               CV.balance = "presences",
                                               CV.strat = "both",
                                               var.import = 1,
                                               metric.eval = c('TSS','ROC'))  #'KAPPA'
AmphiModelOut #check for failed model

cutoff <- get_evaluations(AmphiModelOut)
write.csv(cutoff, file = "cutoff_20240919.csv")

AmphiModelEM <- BIOMOD_EnsembleModeling(bm.mod = AmphiModelOut,
                                                 models.chosen = 'all',
                                                 em.by = 'all',
                                                 em.algo = c(#'EMmean', #'EMcv', 'EMci', 'EMmedian', 'EMca',
                                                 'EMwmean'),
                                                 metric.select = c('TSS'),
                                                 metric.select.thresh = c(0.7),
                                                 metric.eval = c('TSS', 'ROC'),
                                                 var.import = 1,
                                                 EMci.alpha = 0.05, #changed back to 0.05 on 7/25/24
                                                 EMwmean.decay = 'proportional')
AmphiModelEM
get_built_models(AmphiModelEM)
ensemble_kept <- get_kept_models(AmphiModelEM)
write.csv(ensemble_kept, file = "ensemble_kept_20240919.csv")

cutoff_ensemble <- get_evaluations(AmphiModelEM) #summarizes evaluation results
write.csv(cutoff_ensemble, file = "cutoff_ensemble_20240916.csv")

```
[cutoff_20240919.csv](https://github.com/user-attachments/files/17067165/cutoff_20240919.csv)
[cutoff_ensemble_20240916.csv](https://github.com/user-attachments/files/17067166/cutoff_ensemble_20240916.csv)
[ensemble_kept_20240919.csv](https://github.com/user-attachments/files/17067167/ensemble_kept_20240919.csv)

**Comment by MayaGueguen** [comment]
Hello Jackson 👋 

Thank you for posting with lot of details and informations 🙏 

Here is a little recap of what's going on when doing ensemble and evaluating them : 
- first you start with running a bunch of single models : in your case, even with only 3 models, you combine them with pseudo-absence datasets (11) and environmental cross-validation strategy. Which means you "duplicated" your data into 11 datasets, all of them containing all your presences + a selection of a varying number of PA. And then you divided each of these datasets into CV sub-datasets, using a part for calibration and keeping the rest for validation.
- so you obtain as many single models as algorithm x PA x CV (_might not be this number exactly in your case if you provide different PA to each algorithm_), and for each of them, you computing the requested evaluation metrics and select the one threshold that gives you the best evaluation value when transforming probabilities into 0/1 values.
- now you want to keep only the single models that have a minimal evaluation value, and "merge" these kept models into an ensemble model. BUT what you have to keep in mind is that what you will merge are the predictions of these different single models kept, which has no direct link to the evaluation of these new "merged" values.
- This is detailed a bit in slides 41-43 of [this presentation](https://biomodhub.github.io/biomod2/articles/vignette_presentation.html). If you are building an ensemble with the `EMmean` strategy, it means that it will take all your predictions of the (let's say) 5 single models that fullfill the threshold condition, compute in each pixel the average of these 5 models predictions. So you will get a new map of predictions. And you will use this map to compute again your evaluation metrics and find a new threshold to convert predictions to 0/1 values optimizing the evaluation metrics. **What you merge are the predictions, not the evaluation values, which are re-calculated based on the merged predictions**. However, the range of predictions of single models, and the optimized cutoff might have impact on the ensemble cutoff, as discussed below.
- In your case, as you are merging `all` single models together, it means that it cannot keep any longer the splitting between PA and CV datasets : it will take all occurrences, and all points that are included within at least one PA dataset. The model found out that if transforming all predictions > 105 into 1, you manage to predict all your occurrences and absences (PA). As some of your kept single models have cutoff quite low (around 300), it means that in order to match their good evaluation metric, they considered as presences predictions that were quite low. Hence, when doing the average, forcing the ensemble to "move" its optimal cutoff downwards.

Hope it helps, and please do not hesitate if some things still need to be clarified 👀 

Maya

**Comment by jacksonwr5798** [comment]
Thank you for getting back so quickly! Should I be concerned that the cutoffs are so low? With a cutoff that low, my initial thought is that future projections are predicting widespread species distribution when that may not be the case. Does a lower cutoff indicate greater uncertainty in the models? I used this code/method with a few other species and got varying cutoff values for the ensemble (100-700), so I was just curious if that is what was driving some of this.

I am going to rerun a few things based on some of the parameters that you discussed above. I may have additional follow up questions but this was very helpful!

Jackson

**Comment by MayaGueguen** [comment]
Hello Jackson,

Ideally, in such a case, an independant `evaluation` dataset would be the best to try and test your ensemble and see if indeed it leads to too large predictions over the area, or not. But we know that it is hard to have such data :eyes: 

What you could do, is to run also the [BIOMOD_EnsembleModeling](https://biomodhub.github.io/biomod2/reference/BIOMOD_EnsembleModeling.html) function with `EMcv` and `EMci` in the `em.algo` parameter. It might give you some insights on the variability of your predictions, to see if some areas are more prone than other to "disrupt" the model.

Checking variables' importance might also help to check if single models disagree a lot in terms of which variable explain best your occurrences, or to link it with the result of the `EMcv` ensemble (_for example, if one variable is impacting most of the results, and the coefficient of variation identifies areas of high variation in predictions with specific values of this variable, it might help you understanding what's going on_).

Hope it helps !
Maya

**Comment by jacksonwr5798** [comment]
Thank you for these recommendations. Yes, unfortunately we do not have an independent evaluation dataset.

I reran the ensemble to include EMcv and EMci. Those results looked pretty normal to me (see AnaxQue_Current_Ensembleproj_20240923.png). The CV plot indicates very little variance across the study area in the ensemble and the confidence intervals  also show very little uncertainty. 

In checking the variable importance, it was apparent that bio18 was the most important variable and it also had a higher CV value for variable importance compared to other variables included (see summary_var_import_EM_20240923.csv). Would we not expect some degree of variability of in a variable that has higher variable importance? 

I also looked at the response curve plots for EMwmean, EMcv, and EMci. I did notice that the EMcv line was plotting way higher for all of the variables than the EMwmean and EMci lines (see ResponseCurve_ExplVarByRunAlgo_Median_20240923.png and ResponseCurve_ExplVarByRunAlgo_Median_NoCV_20240923.png). Would that explain anything? Note: the red line in ResponseCurve_ExplVarByRunAlgo_Median_20240923.png represents EMcv but in ResponseCurve_ExplVarByRunAlgo_Median_NoCV_20240923.png the red line represents EMwmean.

In addition to those outputs, I was curious if there were significant differences in the distribution generated based on algorithm, PA set, or both PA set and algorithm that were not getting picked in the ensemble when everything merged together. Those results again revealed little variation across algo, PA or PA+algo (see AnaxQue_Current_Ensembleproj_algo_20240926.png, AnaxQue_Current_Ensembleproj_PA_20240926.png, and AnaxQue_Current_Ensembleproj_PAalgo_20240926.png)

Let me know if you have any differing thoughts based on those outputs or if there is something else that we can look into.

![AnaxQue_Current_Ensembleproj_20240923](https://github.com/user-attachments/assets/ac9a1b21-ba1c-4f06-b371-94d4d980eaf8)
[summary_var_import_EM_20240923.csv](https://github.com/user-attachments/files/17153439/summary_var_import_EM_20240923.csv)
![ResponseCurve_ExplVarByRunAlgo_Median_20240923](https://github.com/user-attachments/assets/694675f0-850d-4ba6-9890-8c9cfcc783c1)
![ResponseCurve_ExplVarByRunAlgo_Median_NoCV_20240923](https://github.com/user-attachments/assets/56044ec9-dc86-467e-81a8-d03988dbf2b2)
![AnaxQue_Current_Ensembleproj_algo_20240926](https://github.com/user-attachments/assets/92788ebb-a84d-44be-b33e-65bedd87cbe7)
![AnaxQue_Current_Ensembleproj_PA_20240926](https://github.com/user-attachments/assets/0f0e947d-51ce-49be-8dcd-692cf487f9f3)
![AnaxQue_Current_Ensembleproj_PAalgo_20240926](https://github.com/user-attachments/assets/b803ac9a-5c51-4145-8182-5e806c4825fc)


Thanks,
Jackson

**Comment by MayaGueguen** [comment]
Hello Jackson, 

Indeed, your results seem correct :eyes: 

:warning: maybe try and plot `EMcv` independently of the others as they are not on the same scale.
`EMwmean`, `EMciInf` and `EMciSup` are predictions, while `EMcv` represents a coefficient of variation.

It explains also the scale difference you see on the response curves plot : by default in [bm_PlotResponseCurves](https://biomodhub.github.io/biomod2/reference/bm_PlotResponseCurves.html), predictions are rescaled between 0 and 1 but not the coefficient of variation. You can try and add `on_0_1000 = TRUE` in the parameters if you want to have them alltogether.

Looking at your table file, `bio18` is the variable with the main impact on your predictions, followed by `bio8`.
It means that, if you randomize `bio18`, correlation between your original predictions, and the ones with your variable shuffled will be low. It is normal to see the same pattern for `EMcv`, as by randomizing the variable, you change a lot the predictions, hence you increase the coefficient of variation between predictions.

So it seems that your species is mainly driven by _precipitation of warmest quarter_, and a bit by _temperature of wettest quarter_. 
Main option to see a change in those results would be to add another variable to try and define a bit more your species niche.

Hope it helps,
Maya

**Comment by jacksonwr5798** [comment]
Hey Maya,

I reran them with just emCV, and I looked at the EMcv plot for an ensemble of all of the algorithms together and each algorithm independently. The CV of GBM in particular appears to be more widespread and slightly larger. XGBOOST has the highest cutoff values and also the lowest maximum CV with the greatest uncertainty around the fringes of the species's current range. Are these the observations that you were hoping that I could make from the CV plots? Would these plots help explain the differences in the cutoff values of the different algorithms?

![AnaxQue_Current_Ensembleproj_EMcv_20240927](https://github.com/user-attachments/assets/ba709a05-d4e2-4190-ad27-98d578fd6cc2)
![AnaxQue_Current_Ensembleproj_EMcv_algoGBM_20240927](https://github.com/user-attachments/assets/a6781371-c643-41fd-89e7-b25839b9877c)
![AnaxQue_Current_Ensembleproj_EMcv_algoGLM_20240927](https://github.com/user-attachments/assets/45b513bd-ab17-4690-9725-41b9b82fb2d1)
![AnaxQue_Current_Ensembleproj_EMcv_algoXGBOOST_20240927](https://github.com/user-attachments/assets/ee23c6d2-5ff0-4e70-97fb-c25c2c675794)
[cutoff_ensemble_algo_20240926.csv](https://github.com/user-attachments/files/17170269/cutoff_ensemble_algo_20240926.csv)

I am not necessarily concerned that those variables are the primary drivers since I am predicting the distribution of a frog species. From an ecological prospective, those variables make sense as primary drivers. However, are you suggesting that if I can find another variable that better predicts the niche of the species that might increase the cutoff values for each of the algorithms?

I have been rerunning some things to see how removing some of the algorithms from the ensemble alters the projections and subsequent cutoff values. Ultimately, I would like to avoid output like this. The cutoff is so low that the range change statistics are unrealistic and largely meaningless. 

![AnaxQue_2100ssp585_proj_20240916](https://github.com/user-attachments/assets/1d4e1fc4-039f-4370-aae7-50da1bb6d886)
![CurrentVersus_ssp585_20240916](https://github.com/user-attachments/assets/a06f664a-81cd-4424-9e9f-897779a2118f)

Running the ensemble forecast with just XGBOOST seemed to improve that output, but ideally, I would like to keep at least one boosted model (XGBOOST or GBM) and a linear model (GLM) to maintain the diversity of the algorithms. 

![AnaxQue_ssp585_Ensembleproj_EMwmean_algoXGBOOST_20240927](https://github.com/user-attachments/assets/d89df87e-339d-489e-8ade-7323fb7490ab)
![CurrentVersus_ssp585_XGBOOST_20240927](https://github.com/user-attachments/assets/c8c78a53-73e5-4940-9f8e-fdb06d47a6d8)

Thanks,
Jackson

**Comment by MayaGueguen** [comment]
Hello Jackson,

Indeed, I checked your species, and the model seems to quite agree with where it should be :smile: 

I understand your concern about the range change results with the ensemble model keeping all algorithms, and I don't know between which projection conditions you checked the range change, but keep in mind that it is projecting habitat _suitability_ so the conditions that are OK for the species to be there according to the model you built. So either the species is limited in its range by something else than climate (barriers, land use, other species, etc) or your model is missing environmental conditions to fully understand youe species niche. But if you projected over future conditions, maybe it's just where the species could be ?

Last, in your case, you seem to be better happy with the results of GBM and XGBOOST rather than GLM, but from what I see in your code you sent first, you do a lot of testing for GBM and XGBOOST, and especially you have a number of PA differing for those models (790 vs 10000 for GLM). So maybe try and explore what your results will be with different sizes of PA for all those algorithms ?

Hope it helps,
Maya

**Comment by jacksonwr5798** [comment]
Hey Maya, 

After discussing with another collaborator, I think that we are prepared to accept the model output as it is. The current model is agreeing with where the species current distribution, and we are prepared to discuss the low cutoff values and the vast expansion of habitat suitability in the future projections as an indicatory of uncertainty across the algorithms. We will also just discuss that there is likely a barrier or another explanatory variable that explains this uncertainty. Since we are unsure what the barrier or explanatory variable is, it was not built into the model.

We have been following the PA selection guidelines from Barbet-Massin et al 2012. For GBM and XGBOOST, I selected an equal number of PAs as the number of thinned occurrences with 10 PA sets and for GLM I selected one PA set with 10,000 PA points. We changed the PA settings slightly for one of the species but that was largely because it had fewer occurrences and was projecting into a larger geographic area. The current species has more presences and is projecting into a smaller geographic area than the other species. 

Thank you for your help!

Jackson
